Source/WebCore/Modules/mediasource/SourceBuffer.cpp

```diff
@@ -798,7 +798,7 @@ static PlatformTimeRanges removeSamplesFromTrackBuffer(const DecodeOrderSampleMa
                 additionalErasedRanges.add(previousSample.presentationTime() + previousSample.duration(), erasedStart);
         }
 
-        auto endIterator = trackBuffer.samples.presentationOrder().findSampleStartingOnOrAfterPresentationTime(erasedEnd);
+        auto endIterator = trackBuffer.samples.presentationOrder().findSampleContainingOrAfterPresentationTime(erasedEnd);
```

Comment:
This change fixes high memory usage in apps that do TS -> MP4 transmuxing. The high memory usage is observed in some streams that have a slight overlap in media samples (usually happens on video quality change), when application applies its own logic for samples removal (that looks like sourceBuffer.remove(sourceBuffer.buffered.start(0), ...)). 

Current implementation reports a buffered range that is inconsistent with the actual samples that are still in the buffer (after range removal). The reported range excludes one sample, allowing that sample to "leak". Such a sample holds a reference to the entire appended chunk (due to in-place demuxing in GStreamer MSE backend), but it accounts only for a small size in the buffer. So eviction algorithm doesn't kick in.

Here you can find a mock test that simulates the problem: https://github.com/emutavchi/tests/blob/master/mse/media-source-remove.html. You'll need WebKit with Mock MSE and enable MSE debug logs to see the problem. Here is the output observed without change (range removal part):

```
[WPEWebKit:MediaSource:-] SourceBuffer::remove(29ACFEDF) start = {"value":0}, end = {"value":3.96}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) start = {"value":0}, end = {"value":3.96}          
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":0,"numerator":0,"denominator":1000,"flags":1},"dts":{"value":0,"numerator":0,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"dts":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":2,"numerator":2000,"denominator":1000,"flags":1},"dts":{"value":2,"numerator":2000,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":3,"numerator":3000,"denominator":1000,"flags":1},"dts":{"value":3,"numerator":3000,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removed 256, start = {"value":0,"numerator":0,"denominator":1000,"flags":1}, end = {"value":4,"numerator":4000,"denominator":1000,"flags":1}                                                                                                

[WPEWebKit:MediaSource:-] SourceBuffer::remove(29ACFEDF) start = {"value":4.29}, end = {"value":5}
[WPEWebKit:Media:-] HTMLMediaElement::scheduleUpdatePlayState(29ACFEDF) lambda(), task fired
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) start = {"value":4.29}, end = {"value":5}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":4.29,"numerator":4290,"denominator":1000,"flags":1},"dts":{"value":4.29,"numerator":4290,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":4.62,"numerator":4620,"denominator":1000,"flags":1},"dts":{"value":4.62,"numerator":4620,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":4.95,"numerator":4950,"denominator":1000,"flags":1},"dts":{"value":4.95,"numerator":4950,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removed 192, start = {"value":4.29,"numerator":4290,"denominator":1000,"flags":1}, end = {"value":5.28,"numerator":5280,"denominator":1000,"flags":1} 
```

note that sample at PTS 3.96 "leaked".

here is the output after the fix:
```
[WPEWebKit:MediaSource:-] SourceBuffer::remove(1FCB8181) start = {"value":0}, end = {"value":3.96}                                                                                                         
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) start = {"value":0}, end = {"value":3.96}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":0,"numerator":0,"denominator":1000,"flags":1},"dts":{"value":0,"numerator":0,"denominator":1000,"flags"
:1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"dts":{"value":1,"numerator":1000,"denominator":1000,"
flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":2,"numerator":2000,"denominator":1000,"flags":1},"dts":{"value":2,"numerator":2000,"denominator":1000,"
flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":3,"numerator":3000,"denominator":1000,"flags":1},"dts":{"value":3,"numerator":3000,"denominator":1000,"
flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removed 256, start = {"value":0,"numerator":0,"denominator":1000,"flags":1}, end = {"value":4,"numerator":4000,"denominator":1000,"flag
s":1}

[WPEWebKit:MediaSource:-] SourceBuffer::remove(1FCB8181) start = {"value":4}, end = {"value":5}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) start = {"value":4}, end = {"value":5}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":3.96,"numerator":3960,"denominator":1000,"flags":1},"dts":{"value":3.96,"numerator":3960,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":4.29,"numerator":4290,"denominator":1000,"flags":1},"dts":{"value":4.29,"numerator":4290,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":4.62,"numerator":4620,"denominator":1000,"flags":1},"dts":{"value":4.62,"numerator":4620,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":4.95,"numerator":4950,"denominator":1000,"flags":1},"dts":{"value":4.95,"numerator":4950,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removed 256, start = {"value":3.96,"numerator":3960,"denominator":1000,"flags":1}, end = {"value":5.28,"numerator":5280,"denominator":1
000,"flags":1}
```

Reply:
This result (sample at PTS 3.96 "leaked") is correct:

> Remove all media data, from this [track buffer](https://www.w3.org/TR/media-source/#track-buffer), that contain starting timestamps greater than or equal to start and less than the remove end timestamp.

Calling `sourceBuffer.remove(sourceBuffer.buffered.start(0), 3.96)` should not remove a sample whose presentation timestamp is `3.96` (and a sync sample) because that sample is not "less than the remove end timestamp".

Reply:
Hi @jernoble, thanks for looking into this. The problem is with the buffered range report. The fix doesn't change removal behavior, it only affects the buffered range report. Currently, as you can see, the buffered range starts from 4.29 (after the first removal). This is not correct, 3.96 sample is still there in the buffer.

Reply:
Ah I see! Let me look more into why your proposed change has the effect you're trying to achieve and respond in a bit.

Reply:
Whichever change you decide to go with, please feel free to post it on bugs.webkit.com and CC me; I'd be happy to work with you in getting a fix into the tree.

Reply:
> So there's another problem with this content, in that the sample at pts=3, duration=1 overlaps with the sample at pts=3.96, duration=.33. And that's causing a problem with calculating the end time of the removed range, which is what in turn causes this behavior.

Right, this is what we observe on some live streams in various apps that do TS -> MP4 transmuxing. 

> I don't think that changing the method from startingOnOrAfter to containingOnOrAfter is actually the correct fix for this issue.

The end of the erased range would point either into an empty space or overlapped sample. if it points into an overlapped sample, I think it makes sense to not extend the erased range that would include that sample duration 

> Alternatively, if it is the case that the only way an overlapping sample is allowed to exist in the TrackBuffer is due to a fudge factor

I don't think sample replacement logic should kick in in this case (if I interpreted the MSE spec correctly).

> Whichever change you decide to go with, please feel free to post it on bugs.webkit.com and CC me; I'd be happy to work with you in getting a fix into the tree.

I still think that solution with containingOnOrAfter is good enough to fix the increased memory usage problem that we observe, without having to do a major rework.

Reply:
> > Alternatively, if it is the case that the only way an overlapping sample is allowed to exist in the TrackBuffer is due to a fudge factor
> 
> I don't think sample replacement logic should kick in in this case (if I interpreted the MSE spec correctly).

Except it does. Samples are not allowed to overlap with previously appended samples, but for various reasons (such as bad muxing or variant switching with different timebases) most UAs allow some amount of overlap. If we were strictly adhering to the spec, the sample at PTS=3.96 would evict the sample at PTS=3. 

The code to reconcile removed samples with buffered ranges looks at the end PTS time of those samples in order to define the removed range; if that removed sample (PTS=3) had its duration truncated to 3.96, the reconciliation algorithm would succeed normally.

> > Whichever change you decide to go with, please feel free to post it on bugs.webkit.com and CC me; I'd be happy to work with you in getting a fix into the tree.
> 
> I still think that solution with containingOnOrAfter is good enough to fix the increased memory usage problem that we observe, without having to do a major rework.

The problem with that solution is it does not take into account the amount of overlap. The solution proposed here would allow for situations that absolutely cannot exist, such a removed sample whose timeline is entirely within a non-removed sample, breaking not only the assumption that samples do not overlap, but that samples monotonically increase in both start PTS and end PTS. I would rather not add the concept if a more limited solution (or, IMO, a more correct one) exists.

Reply:
> Except it does.

I must be missing somthing. so, I'm trying to understand how this can happen. Please bear with me.
I see 2 places in the spec that are relevant and could trigger removal of buffered samples:

\# 1
<pre>13. If last decode timestamp for track buffer is unset and presentation timestamp falls within the presentation interval of a coded frame in track buffer, then run the following steps: ... </pre>

The last decoded timestamp is set so we can ignore this case.

\# 2
<pre>14. Remove existing coded frames in track buffer:
If highest end timestamp for track buffer is not set:
    Remove all coded frames from track buffer that have a presentation timestamp greater than or equal to presentation timestamp and less than frame end timestamp.
If highest end timestamp for track buffer is set and less than or equal to presentation timestamp:
   Remove all coded frames from track buffer that have a presentation timestamp greater than or equal to highest end timestamp and less than frame end timestamp.
</pre>

the highest end timestamp is set to 4.0 when an overlapped sample is received. it is higher than pts of "arriving" sample (3.96). so this also doesn't trigger replacement.

I've added additionall logging to trace the processing of coded frames (https://github.com/emutavchi/WebKitForWayland/commit/f11ba3f2fb3124cf9bee3f2c53ceccabc7854e1a).

Here you can find full output: https://gist.github.com/emutavchi/47b4e488156a2018fecb4e533eef2603. 

Processing of sample at 3.96 PTS doesn't trigger removal:
```
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) {"pts":{"value":3.96,"numerator":3960,"denominator":1000,"flags":1},"dts":{"value":3.96,"numerator":3960,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) Last decode time stamp is valid, carry on
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) HighestPresentation timestamp is valid, {"value":4,"numerator":4000,"denominator":1000,"flags":1}
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) HighestPresentation timestamp is valid, no overlap found!
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) erasedSamples is empty
```

so, unless there is something else in the spec that would trigger replacement (and not implementet yet), I conclude that overlapping of media samples is allowed.

Reply:
> > Except it does.
> 
> I must be missing somthing. so, I'm trying to understand how this can happen. Please bear with me. I see 2 places in the spec that are relevant and could trigger removal of buffered samples:
> 
> # 1
> 
> 13. If last decode timestamp for track buffer is unset and presentation timestamp falls within the presentation interval of a coded frame in track buffer, then run the following steps: ... 
> The last decoded timestamp is set so we can ignore this case.

If a container is generating overlapping samples within a coded frame group, it can only be the result of either a bad mux or a demuxing error. The spec has an explicit requirement for monotonically increasing decode timestamps, and an implicit requirement for the sample end times to be well ordered. I'd have to check the ISO BMFF spec, but it should not be the case that samples within a container can be overlapping.

Your test case doesn't reset the (mock) parser between sample `PTS=3` and `PTS=3.96`. Does the in-the-wild behavior you're trying to fix reset the parser (with a new initialization segment) between appends?

Reply:
(As for the Mock test case output and logging, you could do all kind of weird things with the Mock "demure" and get strange behavior; including having non-monotonically increasing decode times, completely overlapping samples, etc. The Mock is just a tool to allow for cross-platform testing, that's all.)

Reply:
I agree that the mock test is oversimplified. I'm sorry if that led to confusion. App does append init segment on quality change right before appending the overlapping samples.  

Here is an example that splits append into two and I think it'd better show the problem: https://raw.githubusercontent.com/emutavchi/tests/master/mse/media-source-remove2.html. I've also included a parser reset (abort) call before appending the second init segment (I'm not sure if the app did that, and I don't have logs from the reproduction case at the moment to confirm).

The behavior didn't change though, there was no sample replaced: https://gist.github.com/emutavchi/4820b0abd59f9e7645663b018d4b826f

Note that PTS and DTS are monotonically increasing.

Reply:
And yes, I agree that this could be an issue with transmuxer on the web app side. but, we cannot fix that.

Reply:
I'd like to recognize that you are indeed correct; the spec does allow for overlapping video samples. I had misremembered; the behavior for audio and text samples is different than for video samples. Personally, I think that's a specification bug, and allowing samples to overlap in the TrackBuffer is confusing and will lead to errors like this one. I would support an addition to MSE that added a "video frame splicing" step similar to the audio and text splice frame algorithm which would simply truncate the duration of the previous sample.  But as the spec stands, that is not required.

That said, I don't think this affects what approach would be the most correct solution to the problem you are trying to solve. Given that overlapping frames can exist in the TrackBuffer, the algorithm's use of `erasedEnd` will inherently cause problems. I still think a more correct solution would be to re-write that section of the algorithm to search the TrackBuffer  using `findSampleStartingAfterPresentationTime()` and `erasedStart`.  Given that samples can overlap, using `findSampleContainingOrAfterPresentationTime()` with `erasedEnd` can still give incorrect results if multiple samples have presentation ranges containing that time.  Because that method uses a binary search algorithm, which sample is found in that scenario is non-deterministic. 

Alternatively, rather than using the sample's duration to calculate the `endTime` of the erased range, the algorithm could use the next remaining sample in PTS order to create the erased range. This would tackle the problem at the correct place, rather than adjusting the erased range to account for overlapping later in the algorithm.

Reply:
I've updated as suggested to use findSampleStartingAfterPresentationTime() and erasedStart. https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/797/commits/880104d32c33cf39f12e42920a4f2bf25df2fd65

Reply:
I've ported the patch and the layout test upstream as https://bugs.webkit.org/show_bug.cgi?id=237780, just in case @jernoble wants to review it there.